### PR TITLE
feat: composite id backend support

### PR DESCRIPTION
**Pull request by yelhouti**
---

Please make sure the below checklist is followed for Pull Requests.

- [X] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [X] Tests are added where necessary
- [X] The JDL part is updated if necessary
- [X] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [X] Documentation is added/updated where necessary
- [X] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by yelhouti** [comment]
fixes #17919

File: generators/database-changelog-liquibase/templates/src/main/resources/config/liquibase/fake-data/table_entity.csv.ejs [hunk 1]
```diff
@@ -59,27 +52,21 @@ for (lineNb = 0; lineNb < entity.fakeDataCount; lineNb++) {
     }
 
     line.push(data);
-    if (field.columnType === 'blob' || field.columnType === 'longblob') {
+      if (field.fieldType === 'byte[]' && field.fieldTypeBlobContent !== 'text') {
```
Comment (Tcharl) [inline generators/database-changelog-liquibase/templates/src/main/resources/config/liquibase/fake-data/table_entity.csv.ejs:55]: There are also derived properties for fieldTypeBytes and fieldTypeBlobContentText

File: generators/entity-client/files-angular.cjs [hunk 2]
```diff
@@ -177,7 +177,7 @@ async function writeAngularFiles() {
     sections: angularFiles,
     rootTemplatesPath: 'angular',
     transform: !application.enableTranslation ? [replaceAngularTranslations] : undefined,
-    context: { ...application, ...entity },
+    context: { ...application, ...entity, entity },
```
Comment (Tcharl) [inline generators/entity-client/files-angular.cjs:180]: Not quite happy with this one, please remove one or the other parameter

File: generators/entity-server/templates/partials/save_template.ejs [hunk 3]
```diff
@@ -43,15 +43,15 @@ if (isUsingMapsId) {
     <%_ resultEntity = persistInstance; _%>
         <%= persistClass %> <%= persistInstance %> = <%= dtoToEntity %>(<%= instanceName %>);
     <%_ if (isUsingMapsId) { _%>
-        <%= mapsIdAssoc.otherEntity.primaryKey.type %> <%= otherEntityName %>Id = <%= instanceName %>.get<%= mapsIdAssoc.relationshipNameCapitalized %>().get<%= primaryKey.nameCapitalized %>();
+        <%= mapsIdAssoc.otherEntity.primaryKey.type %> <%= otherEntityName %>Id = <% if (primaryKey.composite) { %><%= mapper %>.toEntity(<%= instanceName %>)<% } else { %><%= instanceName %><% } %><% if (primaryKey.fields.length === 1) { %>.get<%= mapsIdAssoc.relationshipNameCapitalized %>()<% } %>.get<%= primaryKey.nameCapitalized %>();
```
Comment (Tcharl) [inline generators/entity-server/templates/partials/save_template.ejs:46]: Derived property? i.e. primaryKeyComposite

File: generators/entity-client/files-angular.cjs (hunk 2 above)
Comment (yelhouti) [reply generators/entity-client/files-angular.cjs:180]: @Tcharl, thanks a lot for taking the time to review this.
I need to pass the entity for some recursive code. the ...entity, is used everywhere else to be able to access fields, relationships... directly without `entity.X` . I really think this is the best approach, but maybe you have a better suggestion

File: generators/entity-server/templates/partials/save_template.ejs [hunk 4]
```diff
@@ -43,15 +43,15 @@ if (isUsingMapsId) {
...
         <%= mapsIdRepoInstance %>.findById(<%= otherEntityName %>Id).ifPresent(<%= persistInstance %>::<%_ if (!fluentMethods) { _%>set<%= mapsIdAssoc.relationshipNameCapitalized %> <%_ } else { _%><%= mapsIdAssoc.relationshipName %><%_ } _%>);
     <%_ } _%>
         <%= persistInstance %> = <%= entityInstance %>Repository.save(<%= persistInstance %>);
         <%= returnPrefix %> <%= entityToDto %>(<%= persistInstance %>);
   <%_ } else {
       resultEntity = 'result'; _%>
     <%_ if (isUsingMapsId) { _%>
-        <%= mapsIdAssoc.otherEntity.primaryKey.type %> <%= otherEntityName %>Id = <%= instanceName %>.get<%= mapsIdAssoc.relationshipNameCapitalized %>().get<%= primaryKey.nameCapitalized %>();
+        <%= mapsIdAssoc.otherEntity.primaryKey.type %> <%= otherEntityName %>Id = <%= instanceName %><% if (primaryKey.fields.length === 1) { %>.get<%= mapsIdAssoc.relationshipNameCapitalized %>()<% } %>.get<%= primaryKey.nameCapitalized %>();
```
Comment (Tcharl) [inline generators/entity-server/templates/partials/save_template.ejs:54]: Derived property? i.e. primaryKeyComposite

File: generators/entity-server/templates/partials/update_template.ejs [hunk 5]
```diff
@@ -41,7 +41,7 @@ if (isUsingMapsId) {
         <%= persistInstance %>.setIsPersisted();
     <%_ } _%>
     <%_ if (isUsingMapsId) { _%>
-        <%= mapsIdAssoc.otherEntity.primaryKey.type %> <%= otherEntityName %>Id = <%= instanceName %>.get<%= mapsIdAssoc.relationshipNameCapitalized %>().get<%= primaryKey.nameCapitalized %>();
+        <%= mapsIdAssoc.otherEntity.primaryKey.type %> <%= otherEntityName %>Id = <% if (primaryKey.composite) { %><%= mapper %>.toEntity(<%= instanceName %>)<% } else { %><%= instanceName %><% } %><% if (primaryKey.fields.length === 1) { %>.get<%= mapsIdAssoc.relationshipNameCapitalized %>()<% } %>.get<%= primaryKey.nameCapitalized %>();
```
Comment (Tcharl) [inline generators/entity-server/templates/partials/update_template.ejs:44]: Derived property? i.e. primaryKeyComposite

File: generators/entity-server/templates/partials/update_template.ejs [hunk 6]
```diff
@@ -53,7 +53,7 @@ if (isUsingMapsId) {
   <%_ } else {
       resultEntity = 'result'; _%>
     <%_ if (isUsingMapsId) { _%>
-        <%= mapsIdAssoc.otherEntity.primaryKey.type %> <%= otherEntityName %>Id = <%= instanceName %>.get<%= mapsIdAssoc.relationshipNameCapitalized %>().get<%= primaryKey.nameCapitalized %>();
+        <%= mapsIdAssoc.otherEntity.primaryKey.type %> <%= otherEntityName %>Id = <%= instanceName %><% if (primaryKey.fields.length === 1) { %>.get<%= mapsIdAssoc.relationshipNameCapitalized %>()<% } %>.get<%= primaryKey.nameCapitalized %>();
```
Comment (Tcharl) [inline generators/entity-server/templates/partials/update_template.ejs:56]: Derived property? i.e. primaryKeyComposite

File: generators/entity-server/templates/src/main/java/package/domain/Entity.java.jhi.ejs [hunk 7]
```diff
@@ -102,13 +102,13 @@ public class <%= persistClass %> <&- fragments.extendsSection() -&>implements Se
...
 <%_ if (!embedded && primaryKey.composite) { _%>
-    @Id
+    @EmbeddedId
     private <%= primaryKey.type %> <%= primaryKey.name %>;
 
 <%_ } _%>
 <&- fragments.classStaticFieldsSection() -&>
 <&- fragments.classFieldsSection() -&>
-<%_ for (const field of fields.filter(field => !field.javaInherited && !field.transient && (embedded || !field.id || !primaryKey.composite))) { _%>
+<%_ for (const field of fields.filter(field => !field.javaInherited && !field.transient && (embedded || !(field.id && primaryKey.composite && primaryKey.fields.length === 1)))) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/domain/Entity.java.jhi.ejs:111]: primaryKey.composite

File: generators/entity-server/templates/src/main/java/package/domain/Entity.java.jhi.ejs [hunk 8]
```diff
@@ -176,7 +176,24 @@ for (relationship of relationships.filter(relationship => !relationship.embedded
...
+        this.set<%= primaryKey.nameCapitalized %>(<%= primaryKey.name %>);
+        return this;
+    }
+  <%_ } _%>
+
+    public void set<%= primaryKey.nameCapitalized %>(<%= primaryKey.type %> <%= primaryKey.name %>) {
+        this.<%= primaryKey.name %> = <%= primaryKey.name %>;
+    }
+<%_ } _%>
+<%_ for (const field of fields.filter(field => !field.transient && (embedded || !(field.id && primaryKey.composite && primaryKey.fields.length === 1)))) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/domain/Entity.java.jhi.ejs:196]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/service/mapper/EntityMapper.java.ejs [hunk 9]
```diff
@@ -80,10 +79,16 @@ public interface <%= entityClass %>Mapper extends EntityMapper<<%= dtoClass %>,
   <%_ } _%>
 
   <%_ var renMapAnotDto = false;  //Render Mapping Annotation during DTO to Entity conversion? _%>
-  <%_ if(primaryKey.ids.length > 1) { _%>
+  <%_ if(primaryKey.fields.length > 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/service/mapper/EntityMapper.java.ejs:82]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 10]
```diff
@@ -24,6 +24,11 @@ _%>
 <%_ if (!dtoMapstruct || serviceNo) { _%>
 import <%= entityAbsolutePackage %>.domain.<%= persistClass %>;
 <%_ } _%>
+<%_ if (primaryKey.fields.length > 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:27]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 11]
```diff
@@ -136,6 +141,10 @@ import static org.elasticsearch.index.query.QueryBuilders.*;
 public class <%= entityClass %>Resource {
 
     private final Logger log = LoggerFactory.getLogger(<%= entityClass %>Resource.class);
+
+<%_ if(primaryKey.fields.length > 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:145]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 12]
```diff
@@ -165,9 +175,27 @@ public class <%= entityClass %>Resource {
     @PostMapping("/<%= entityApiUrl %>")
     public <% if (reactive) { %>Mono<<% } %>ResponseEntity<<%= instanceType %>><% if (reactive) { %>><% } %> create<%= entityClass %>(<% if (validation) { %>@Valid <% } %>@RequestBody <%= instanceType %> <%= instanceName %>) throws URISyntaxException {
         log.debug("REST request to save <%= entityClass %> : {}", <%= instanceName %>);
+  <%_ if (primaryKey.fields.length === 1 && primaryKey.autoGenerate) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:178]: primaryKey.containsUniqueField?

Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:178]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 13]
```diff
@@ -150,6 +159,7 @@ public class <%= entityClass %>Resource {
     <%_
     const instanceType = restClass;
     const instanceName = restInstance;
+    const apiUrlPath = (key) => primaryKey.fields.length === 1 ? `${key}.get${primaryKey.nameCapitalized}()` : primaryKey.fields.map(f => `"${f.fieldName}=" + ${key}.${f.path.map(n => `get${this._.upperFirst(n)}()`).join('.')}`).join(' + ";" + ');
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:162]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs (hunk 11 above)
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:145]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 14]
```diff
@@ -165,9 +175,27 @@ public class <%= entityClass %>Resource {
...
     public <% if (reactive) { %>Mono<<% } %>ResponseEntity<<%= instanceType %>><% if (reactive) { %>><% } %> create<%= entityClass %>(<% if (validation) { %>@Valid <% } %>@RequestBody <%= instanceType %> <%= instanceName %>) throws URISyntaxException {
         log.debug("REST request to save <%= entityClass %> : {}", <%= instanceName %>);
+  <%_ if (primaryKey.fields.length === 1 && primaryKey.autoGenerate) { _%>
         if (<%= instanceName %>.get<%= primaryKey.nameCapitalized %>() != null) {
             throw new BadRequestAlertException("A new <%= entityInstance %> cannot already have an ID", ENTITY_NAME, "idexists");
         }
+  <%_ } _%>
+<%_ // TODO handle these cases when reactive (using switchIfEmpty flatMap...
+if (!reactive) { _%>
+    <%_ if(primaryKey.fields.length === 1 && !primaryKey.autoGenerate) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:185]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 15]
```diff
@@ -165,9 +175,27 @@ public class <%= entityClass %>Resource {
...
+        if (<%= entityInstance %>Repository.existsById(<%= instanceName %>.get<%= primaryKey.nameCapitalized %>())) {
+            throw new BadRequestAlertException("This <%= entityInstance %> already exists", ENTITY_NAME, "idduplicate");
+        }
+    <%_ } else if (primaryKey.composite) { _%>
+        <%= primaryKey.type %> id = <%= entityInstance %>Mapper.toEntity(<%= instanceName %>).getId();
+        if (id == null) {
+            throw new BadRequestAlertException("Invalid id", ENTITY_NAME, "idnull");
+        }
+        if (<%= entityInstance %>Service.findOne(id).isPresent()) {
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:194]: prefer use map().orElseGet()

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 16]
```diff
@@ -218,14 +250,27 @@ public class <%= entityClass %>Resource {
      */
     @PutMapping("/<%= entityApiUrl %>/{<%= primaryKey.name %>}")
     public <% if (reactive) { %>Mono<<% } %>ResponseEntity<<%= instanceType %>><% if (reactive) { %>><% } %> update<%= entityClass %>(
+<%_ if (primaryKey.fields.length === 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:253]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 17]
```diff
@@ -218,14 +250,27 @@ public class <%= entityClass %>Resource {
...
         @PathVariable(value = "<%= primaryKey.name %>", required = false) final <%= primaryKey.type %> <%= primaryKey.name %>,
+<%_ } else { _%>
+        @MatrixVariable(pathVar = "id") Map<String, String> idMap,
+<%_ } _%>
         <% if (validation) { %>@Valid <% } %>@RequestBody <%= instanceType %> <%= instanceName %>
     ) throws URISyntaxException {
+<%_ if(primaryKey.fields.length > 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:260]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 18]
```diff
@@ -218,14 +250,27 @@ public class <%= entityClass %>Resource {
...
+        final <%= primaryKey.type %> id = mapper.convertValue(idMap, <%= primaryKey.type %>.class);
+<%_ } _%>
         log.debug("REST request to update <%= entityClass %> : {}, {}", <%= primaryKey.name %>, <%= instanceName %>);
+<%_ if(primaryKey.fields.length === 1) { _%>
         if (<%= instanceName %>.get<%= primaryKey.nameCapitalized %>() == null) {
             throw new BadRequestAlertException("Invalid id", ENTITY_NAME, "idnull");
         }
+<%_ } _%>
+<%_ if(primaryKey.fields.length === 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:269]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 19]
```diff
@@ -274,7 +319,11 @@ public class <%= entityClass %>Resource {
     /**
      * {@code PATCH  /<%= entityApiUrl %>/:<%= primaryKey.name %>} : Partial updates given fields of an existing <%= entityInstance %>, field will ignore if it is null
      *
+<%_ if(primaryKey.fields.length === 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:322]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 20]
```diff
@@ -284,13 +333,26 @@ public class <%= entityClass %>Resource {
      */
     @PatchMapping(value = "/<%= entityApiUrl %>/{<%= primaryKey.name %>}", consumes = {"application/json", "application/merge-patch+json"})
     public <% if (reactive) { %>Mono<<% } %>ResponseEntity<<%= instanceType %>><% if (reactive) { %>><% } %> partialUpdate<%= entityClass %>(
-    @PathVariable(value = "<%= primaryKey.name %>", required = false) final <%= primaryKey.type %> <%= primaryKey.name %>,
+<%_ if (primaryKey.fields.length === 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:336]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 21]
```diff
@@ -284,13 +333,26 @@ public class <%= entityClass %>Resource {
...
+        @PathVariable(value = "<%= primaryKey.name %>", required = false) final <%= primaryKey.type %> <%= primaryKey.name %>,
+<%_ } else { _%>
+        @MatrixVariable(pathVar = "id") Map<String, String> idMap,
+<%_ } _%>
     <% if (validation) { %>@NotNull <% } %>@RequestBody <%= instanceType %> <%= instanceName %>) throws URISyntaxException {
+<%_ if (primaryKey.fields.length > 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:342]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 22]
```diff
@@ -284,13 +333,26 @@ public class <%= entityClass %>Resource {
...
+        final <%= primaryKey.type %> id = mapper.convertValue(idMap, <%= primaryKey.type %>.class);
+<%_ } _%>
         log.debug("REST request to partial update <%= entityClass %> partially : {}, {}", <%= primaryKey.name %>, <%= instanceName %>);
+<%_ if(primaryKey.fields.length === 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:346]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 23]
```diff
@@ -284,13 +333,26 @@ public class <%= entityClass %>Resource {
...
         if (<%= instanceName %>.get<%= primaryKey.nameCapitalized %>() == null) {
             throw new BadRequestAlertException("Invalid id", ENTITY_NAME, "idnull");
         }
+<%_ } _%>
+<%_ if(primaryKey.fields.length === 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:351]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 24]
```diff
@@ -386,14 +448,21 @@ public class <%= entityClass %>Resource {
     /**
      * {@code GET  /<%= entityApiUrl %>/:id} : get the "id" <%= entityInstance %>.
      *
+<%_ if(primaryKey.fields.length === 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:451]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 25]
```diff
@@ -386,14 +448,21 @@ public class <%= entityClass %>Resource {
...
+     * @param idMap a Map representation of the id of the <%= instanceName %> to retrieve.
+<%_ } _%>
      * @return the {@link ResponseEntity} with status {@code 200 (OK)} and with body the <%= instanceName %>, or with status {@code 404 (Not Found)}.
      */
     @GetMapping("/<%= entityApiUrl %>/{id}")
 <%_ if (databaseTypeSql && isUsingMapsId && !viaService) { _%>
     @Transactional(readOnly = true)
 <%_ } _%>
-    public <% if (reactive) { %>Mono<<% } %>ResponseEntity<<%= instanceType %>><% if (reactive) { %>><% } %> get<%= entityClass %>(@PathVariable <%= primaryKey.type %> id) {
+    public <% if (reactive) { %>Mono<<% } %>ResponseEntity<<%= instanceType %>><% if (reactive) { %>><% } %> get<%= entityClass %>(<% if (primaryKey.fields.length === 1) { %>@PathVariable <%= primaryKey.type %> id<% } else { %>@MatrixVariable(pathVar = "id") Map<String, String> idMap<% } %>) {
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:462]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 26]
```diff
@@ -386,14 +448,21 @@ public class <%= entityClass %>Resource {
...
+<%_ if (primaryKey.fields.length > 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:463]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 27]
```diff
@@ -402,11 +471,18 @@ public class <%= entityClass %>Resource {
     /**
      * {@code DELETE  /<%= entityApiUrl %>/:id} : delete the "id" <%= entityInstance %>.
      *
+<%_ if(primaryKey.fields.length === 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:474]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 28]
```diff
@@ -402,11 +471,18 @@ public class <%= entityClass %>Resource {
...
      * @param id the id of the <%= instanceName %> to delete.
+<%_ } else { _%>
+     * @param idMap a Map representation of the id of the <%= instanceName %> to delete.
+<%_ } _%>
      * @return the {@link ResponseEntity} with status {@code 204 (NO_CONTENT)}.
      */
     @DeleteMapping("/<%= entityApiUrl %>/{id}")
-    public <% if (reactive) { %>Mono<<% } %>ResponseEntity<Void><% if (reactive) { %>><% } %> delete<%= entityClass %>(@PathVariable <%= primaryKey.type %> id) {
+    public <% if (reactive) { %>Mono<<% } %>ResponseEntity<Void><% if (reactive) { %>><% } %> delete<%= entityClass %>(<% if (primaryKey.fields.length === 1) { %>@PathVariable <%= primaryKey.type %> id<% } else { %>@MatrixVariable(pathVar = "id") Map<String, String> idMap<% } %>) {
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:482]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 29]
```diff
@@ -402,11 +471,18 @@ public class <%= entityClass %>Resource {
...
+<%_ if (primaryKey.fields.length > 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:483]: !primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 30]
```diff
@@ -420,7 +496,7 @@ public class <%= entityClass %>Resource {
         );
     <%_ } _%>
   <%_ } else { _%>
-        return ResponseEntity.noContent().headers(HeaderUtil.createEntityDeletionAlert(applicationName, <%= enableTranslation %>, ENTITY_NAME, id<% if (!primaryKey.fields[0].fieldTypeString) { %>.toString()<% } %>)).build();
+        return ResponseEntity.noContent().headers(HeaderUtil.createEntityDeletionAlert(applicationName, <%= enableTranslation %>, ENTITY_NAME, id<% if (primaryKey.type !== 'String') { %>.toString()<% } %>)).build();
```
Comment (Tcharl) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs:499]: !primaryKeyTypeString

File: generators/entity-server/templates/src/test/java/package/service/dto/EntityDTOTest.java.ejs [hunk 31]
```diff
@@ -21,33 +21,49 @@ package <%= entityAbsolutePackage %>.service.dto;
...
 import java.util.UUID;
 <%_ } _%>
+<%_
+const initDTO = (dtoEntity, entityNbr) => {
+_%>
+    <%= dtoEntity.dtoClass %> <%= dtoEntity.dtoInstance %><%= entityNbr %> = new <%= dtoEntity.dtoClass %>();
+    <%_ dtoEntity.fields.filter(f => f.id).forEach(f => { _%>
+    <%= dtoEntity.dtoInstance + entityNbr %>.set<%= f.fieldNameCapitalized %>(<%- this.getPrimaryKeyValue(f.fieldType, databaseType, entityNbr) %>);
+    <%_ }) _%>
+    <%_ dtoEntity.relationships.filter(r => r.id && !(r.relationshipType === 'one-to-one' && !dtoEntity.primaryKey.composite)).forEach(r => { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/test/java/package/service/dto/EntityDTOTest.java.ejs:34]: relationshipTypeOneToOne

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs [hunk 32]
```diff
@@ -55,13 +55,24 @@ let transactionalAnnotation = '';
 if (databaseTypeSql && !reactive) {
   transactionalAnnotation = '\n    @Transactional';
 }
-
+const persistInstanceUrlId = primaryKey.fields.length === 1 ?
```
Comment (Tcharl) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:58]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs [hunk 33]
```diff
@@ -55,13 +55,24 @@ let transactionalAnnotation = '';
...
+    `${persistInstance}.get${primaryKey.nameCapitalized}()` :
+    primaryKey.fields.map(field => `"${field.fieldName}=" + ${persistInstance}.getId().get${field.fieldNameCapitalized}()`).join(' + ";" + ');
+let generatedUrlId;
+if(primaryKey.fields.length === 1) {
```
Comment (Tcharl) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:62]: primaryKey.containsUniqueField?

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs [hunk 34]
```diff
@@ -190,7 +201,7 @@ import java.util.stream.Stream;
   <%_ } _%>
 <%_ } _%>
 import java.util.List;
-<%_ if (fieldsContainUUID || primaryKey.typeString || otherEntityPrimaryKeyTypesIncludesUUID) { _%>
+<%_ if (fieldsContainUUID || primaryKey.typeString || primaryKey.fields.map(f => f.fieldType).some(t => ['String', 'UUID'].includes(t)) || otherEntityPrimaryKeyTypesIncludesUUID) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:204]: Update fieldsContainsUUID and add derived primaryKeyContainsString (not sure primaryKeyTypeString is useful anymore)

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs [hunk 35]
```diff
@@ -716,7 +732,9 @@ _%>
   <%_ } _%>
     }
 
-  <%_ if (databaseTypeSql && isUsingMapsId) { _%>
+  <%# this is not doable with composite keys, without drastically chaging the logic, since we would need to read the id from url,
+   set it in entity instead of validating it, then using the relationnships to update the DB, and it might not work since we usable insertable/updatable false %>
+  <%_ if (databaseTypeSql && isUsingMapsId && primaryKey.fields.length === 1) { _%>
```
Comment (Tcharl) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:737]: primaryKey.containsUniqueField?

**Comment by mraible** [comment]
Conflicts need to be resolved.

**Comment by DanielFran** [comment]
@yelhouti Are you still committed to working on this feature?

**Comment by yelhouti** [comment]
Yes, If anyone is willing to Review and merge.

**Comment by DanielFran** [comment]
@yelhouti you need ist to fix the conflicts

**Comment by yelhouti** [comment]
@DanielFran I did that a few times and as you can imagine, with this big of PR it takes time. As soon as someone volontiers for merging I can do it again. I can also pay one of the maintainers if they want for their time to implement this correctly for the backend. @mshima interested ?

**Comment by yelhouti** [comment]
@DanielFran I fixed the conflicts (if you could please have a look) I still have some small tests to fix ... I m willing to put more time if you are willing to merge

**Comment by yelhouti** [comment]
@mshima thanks a lot man for the work you did on vscode to generate integration tests locally. It was always a PITA to know which test should be run how. Now it all works great and I can fix tests 10 times faster. Maybe this should be documented for others in [CONTRIBUTING.md](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md#running-integration-tests-locally) .

**Comment by yelhouti** [comment]
@mshima @DanielFran finally tests are passing here and in my compositeKey projects. Could anyone please merge this

**Review by mshima — CHANGES_REQUESTED** [review]
Entity suffix is not correct.
I need to do a deeper review.

**Comment by yelhouti** [comment]
@mshima sure, if you can point me to what to fix, I will gladly do

**Comment by mshima** [comment]
At least it looks wrong at generated sample.
IMG_0276

**Comment by yelhouti** [comment]
> At least it looks wrong at generated sample. IMG_0276

When using MapsId, the id of the parent and the child are the same. Put I can change this, if you want, hopefully it won't break the composite with multi level... I must admit I don't remember  why i changed it

**Comment by yelhouti** [comment]
Guess, I need to rebase again ...

**Review by mshima — CHANGES_REQUESTED** [review]
We should avoid complicating templates.

This is a partial review, would be much better if split into multiple PRs.

**Pull request closed without merging**